### PR TITLE
net: lwm2m: add certificate support

### PR DESCRIPTION
**Pull request by RomainPelletant**
Add option for x509 certificate support according to Security object 1.1
3: client's certificate
4: server's certificate (CA)
5: client's pkey
Add optionnal CA check to be used easily with leshan default certificate
Issue #54376

Signed-off-by: romain pelletant <romainp@kickmaker.net>

### CODE REVIEW

File: subsys/net/lib/lwm2m/Kconfig [hunk 1]
```diff
@@ -43,6 +43,28 @@ config LWM2M_SHELL
...
+
+endchoice
+
+config LWM2M_CHECK_SERVER_CERTIFICATE
+	bool "Check returned CA server certificate"
+	depends on LWM2M_DTLS_SECURITY_CERTIFICATE
+	default y
+	help
+	  Check CA of server's certificate. Useful to test Leshan.
+	  Do not use in production mode.
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:66]: I think this is a bit confusing, on one hand we have a config that should be unquestionably enabled in any case other than debugging, on other, the help message says `Do not use in production mode.`.

I think it'd be more intuitive if the logic here was reversed, i. e. provide a config, that if enabled, disabler server certificate verification.

**Comment by SeppoTakalo** [comment]
Actually.. do we need to touch the engine at all?

Can you study what we did on the nRF Connect SDK: https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c

The engine already supports X509 as long as those are provisioned to the given security tag. So these Kconfigs might not bring any value, other than hard-code the choice.

Then, also the `context->load_credentials()` should probably be used instead of doing the stuff in the engine.

**Comment by RomainPelletant** [comment]
@SeppoTakalo After havin a look at lwm2m_security.c, it does not really make sense to implement it in lwm2m_engine.c.

As you suggested, we could only add the following modifications:

- Kconfig : Add LWM2M_SECURITY_IDENTITY_SIZE 
- Kconfig : Increase LWM2M_SECURITY_KEY_SIZE up to 1024
- Kconfig : Add LWM2M_SKIP_SERVER_CERTIFICATE but with reverse logic according to rlubos comment
- lwm2m_engine.c : Add skip server certificate to ease leshan demo tests

**Comment by SeppoTakalo** [comment]
I did some trial...

I don't think you need to touch the engine that much. I was able to achieve X509 support just by providing the `load_credentials()` callback like this:

```c

char server_certificate[] =
"-----BEGIN CERTIFICATE-----\n"
"..."
"-----END CERTIFICATE-----\n";

char private_key[] =
"-----BEGIN EC PRIVATE KEY-----\n"
"..."
"-----END EC PRIVATE KEY-----\n";

char certificate[] =
"-----BEGIN CERTIFICATE-----\n"
"..."
"-----END CERTIFICATE-----\n";

char hostname[] = "localhost";

int load_credentials(struct lwm2m_ctx *client_ctx)
{
        int ret;

        client_ctx->tls_tag = CONFIG_LWM2M_APP_TLS_TAG;

        ret = tls_credential_add(CONFIG_LWM2M_APP_TLS_TAG, TLS_CREDENTIAL_CA_CERTIFICATE,
                         server_certificate, sizeof(server_certificate));
        if (ret) {
                LOG_ERR("tls_credential_add(CA_CERTIFICATE) failed %d", ret);
                return ret;
        }

        ret = tls_credential_add(CONFIG_LWM2M_APP_TLS_TAG, TLS_CREDENTIAL_SERVER_CERTIFICATE,
                         certificate, sizeof(certificate));
        if (ret) {
                LOG_ERR("tls_credential_add(SERVER_CERTIFICATE) failed %d", ret);
                return ret;
        }

        ret = tls_credential_add(CONFIG_LWM2M_APP_TLS_TAG, TLS_CREDENTIAL_PRIVATE_KEY,
                         private_key, sizeof(private_key));
        if (ret) {
                LOG_ERR("tls_credential_add(PRIVATE_KEY) failed %d", ret);
                return ret;
        }

        client_ctx->desthostname = hostname;
        client_ctx->desthostnamelen = sizeof(hostname);
        client_ctx->hostname_verify = true;

        LOG_INF("load_credentials() OK");
        return 0;
}
```
Notice that I can use just one security tag, like it is currently in Kconfig. Then some trial&error with MbedTLS Kconfig values and I got it working.

Of course, this is not a complete example, as I'm not using anything from security object. However, I think it is wrong to implement PSK like it is now in the `lwm2m_socket_start()`.
I think we should split out the PSK to `load_psk(...)` and copy&paste my code to provide `load_certificate()` and implement it so that it loads from the selected security object, instead of those `char *` pointers.

Then, only thing in the engine should be like `if (ctx->load_credentials()`   then use that, otherwise try to read from the selected security object which mode it is and call either `load_psk()` or `load_x509()`. Rest would be just Kconfig values, maybe in a separate overlay.

**Comment by SeppoTakalo** [comment]
@RomainPelletant 
I had a second thought on this.
Also, I would like to hear @rlubos feedback on this.

I have three suggestions that we should do:
1. Socket should be opened before we call `ctx->load_credentials()`. This would allow application to override socket options. So move the call to `lwm2m_open_socket()` before those calls.
2. Engine should provide default implementation of `load_credentials()`, easiest would be to provide one wrapper that checks if the mode is PSK or X509, then call `load_credentials_psk()` or `load_credentials_x509()`.
3. All security related socket options should be moved to `load_credentials()`. So when you override, you have a full control of the socket options. Move the existing ones into the default `load_credentials()`.

When those are done, I think we could consider dropping the `desthostname`, `hostname_verify` and `use_dtls` flags from the context. After all, that information should be in security object, not hard coded into the context.

**Comment by rlubos** [comment]
> When those are done, I think we could consider dropping the desthostname, hostname_verify and use_dtls flags from the context. After all, that information should be in security object, not hard coded into the context.

I agree, that if we're into more serious refactoring, that should be the way to go. In theory, DTLS configuration could change during bootstrap, so nothing should really be hardcoded on the application side. All information on what security type to use, and what credentials, should be read from the security context.

> Socket should be opened before we call ctx->load_credentials(). This would allow application to override socket options. So move the call to lwm2m_open_socket() before those calls.

I think this could be done regardless, but I just wonder if there's really a need to override the security configuration in the basic scenario. There's no really a socket option to differentiate between PSK/certificates, all the difference comes from the credential type stored under the security tag. The security tag the credentials were stored under could for example be returned by the `load_credentials()` function. But maybe there's some use-case I don't see though.

**Comment by RomainPelletant** [comment]
As @SeppoTakalo and @rlubos said, x509 is more related to configuration and commonly changed during bootstrap session.

After PSK connection to bootstrap, it send back the x509 credentials over the wire on the device. The limitation I encountered is the block-wise support because (client certificate, private key and server CA) are greater than 1024 bytes.
So in the meanwhile (block-wise implemented) I don't think it is interesting to implement it as fully supported.

If you agree, I would implement the solution @SeppoTakalo shared as low priority task.

> @RomainPelletant I had a second thought on this. Also, I would like to hear @rlubos feedback on this.
> 
> I have three suggestions that we should do:
> 
> 1. Socket should be opened before we call `ctx->load_credentials()`. This would allow application to override socket options. So move the call to `lwm2m_open_socket()` before those calls.
> 2. Engine should provide default implementation of `load_credentials()`, easiest would be to provide one wrapper that checks if the mode is PSK or X509, then call `load_credentials_psk()` or `load_credentials_x509()`.
> 3. All security related socket options should be moved to `load_credentials()`. So when you override, you have a full control of the socket options. Move the existing ones into the default `load_credentials()`.
> 
> When those are done, I think we could consider dropping the `desthostname`, `hostname_verify` and `use_dtls` flags from the context. After all, that information should be in security object, not hard coded into the context.

I would prefer implement the EST security client as it should not be impacted by block-wise transfer.

**Comment by MaureenHelm** [comment]
@SeppoTakalo please revisit

**Comment by MaureenHelm** [comment]
@rlubos will you please take a look?

**Comment by rlubos** [comment]
I had an impression the PR is pending some changes:
> I would implement the solution @SeppoTakalo shared as low priority task.

**Comment by MaureenHelm** [comment]
> I have submitted an alternative approach #59019

Can this PR be closed since the alternative approach was merged?

**Pull request closed without merging**